Big QP parameter refactor; validate param :types for Cards

camsaul opened the pull request:
Part of https://github.com/metabase/metaboat/issues/144

The main purpose of this PR is to enforce a new constraint that all `:parameters` passed in to Card query endpoints match up with the template tag parameters for the Card's query (only native queries can have template tag parameters). This means you can no longer pass arbitrary MBQL parameters when running a Card query. Parameters must match existing parameters and have allowed types, e.g. you cannot pass a `string/!=` parameter when the template tag specifies a `:widget-type` of `:string/=` (for security reasons explained in the comments in the PR).

As part of this project I undertook a few important refactors:

- Code for running queries in the context of a specific Card or Dashboard+Card has been moved to new namespaces `metabase.query-processor.card` and `metabase.query-processor.dashboard` respectively. Previously they were in `metabase.api.card` and `metabase.api.dashboard`.
- Schemas for native query `:template-tags` and for `:parameters` are now finally moved into `metabase.mbql.schema` itself. This actually involved consolidating dispersed information from four or five different namespaces. I tried my best to document all the possible values of these
- (As part of the schema above) finally coming up with a list of the allowed parameter types and widget types -- previously this list didn't exist anywhere. I had to update some tests because they were doing things that didn't make sense or were missing some keys they should have been passing
- The backend parameters code was kinda fragile and every time I changed something, something else would break in an e2e test. I added close to 600 LoC of new backend tests as part of this PR

camsaul committed:
Refactor: move Card and Dashboard QP code into their own qp.* namespaces

camsaul committed:
Disable extra validation for now so a million tests don't fail

camsaul committed:
WIP

camsaul committed:
Merge branch 'master' into validate-card-parameters

camsaul committed:
Validate template tag :parameters in query in context of a Card

camsaul committed:
Fixes

camsaul committed:
Merge branch 'master' into validate-card-parameters

camsaul committed:
Disable strict validation for now

camsaul committed:
Test fixes [WIP]

camsaul committed:
Make the parameter type schema a little more forgiving for now

camsaul committed:
Tests & test fixes :wrench:

camsaul committed:
More test fixes :wrench:

camsaul committed:
Merge branch 'master' into validate-card-parameters

camsaul committed:
1. Need more tests
2. Need to actually validate stuff

camsaul committed:
More test fixes. :wrench:

camsaul committed:
Test fixes (again)

camsaul committed:
Test fix :wrench:

dpsutton commented on test/metabase/api/card_test.clj:
```diff
@@ -11,17 +11,28 @@
             [metabase.http-client :as http]
-            [metabase.models :refer [Card CardFavorite Collection Dashboard Database ModerationReview
-                                     Pulse PulseCard PulseChannel PulseChannelRecipient Table ViewLog]]
+            [metabase.models
```
This change is probably not desired. Auto format got a bit hungry

camsaul replied:
good catch. I'll fix

dpsutton commented on test/metabase/pulse/render/body_test.clj:
```diff
@@ -7,6 +7,12 @@
             [metabase.pulse.render.test-util :as render.tu]
             [schema.core :as s]))
 
+(use-fixtures :each
```
this is weird. Did this not get merged already?

dpsutton replied:
oh i see it was added to js-svg-test and copied to body-test here

camsaul replied:
yeah, I ran into it again in this namespace as well

dpsutton reviewed (approved):
This is fantastic. Nice!

camsaul committed:
Some test fixes / PR feedback

daltojohnso committed:
Disallow native queries with a tag widget-type of "none"

Template tags with a widget-type that is undefined, null, or "none" now
cause the query's isRunnable method to return false. Existing questions
that have this defect won't be runnable until they are resaved with a
set widget-type.

daltojohnso committed:
Fix prettier error

daltojohnso committed:
add snippet and card types to validation pass

camsaul committed:
Merge branch 'master' into validate-card-parameters

camsaul committed:
Make sure template tag map keys + `:names` agree + test fixes

camsaul committed:
Have MBQL normalization reconcile template tags map key and :name

camsaul committed:
Test fix :wrench:

camsaul committed:
Fix tests for Cljs

camsaul committed:
Fix Mongo tests.

camsaul committed:
Allow passing :category parameters for :text/:number/:date for now.

camsaul committed:
Dashboard subscriptions should use qp.dashboard code for executing

camsaul committed:
Make sure Dashboard QP parameter resolution code merges in default values

camsaul committed:
Add a test for sending a test Dashboard subscription with default params

camsaul committed:
Prettier

camsaul committed:
If both Dashboard and Card have default param value, prefer Card's default

camsaul committed:
Test fix :wrench:

camsaul committed:
More tests and more fixes :cry:

camsaul commented:
Presto JDBC failure was unrelated and fixed upstream by #19256. Everything else is green -- merging
